### PR TITLE
Pbindef argument order

### PR DESCRIPTION
**Pull request by telephon**
This will insert keys instead of appending them.

### CODE REVIEW

File: testsuite/classlibrary/TestPatternProxy.sc [hunk 1]
```diff
@@ -366,4 +366,19 @@ TestPatternProxy : UnitTest {
...
 	}
+
+	test_pbindef_arg_order {
+		var pairs;
+
+		Pbindef(\test).clear;
+		Pbindef(\test, \a, 1, \b, 2, \c, 3);
+
+		pairs = Pbindef(\test).source.pairs;
+		this.assertEquals(pairs[0,2..], [\a, \b, \c], "Pbindef key order should correspond to code");
```
Comment (brianlheim) [inline testsuite/classlibrary/TestPatternProxy.sc]: this assert and the line above are unnecessary, probably best to remove them. this test should only ensure that adding new keys inserts, not appends; any other behavior should be guaranteed by a separate test

**Comment by brianlheim** [comment]
nice idea, thanks!

File: testsuite/classlibrary/TestPatternProxy.sc (hunk 1 above)
Comment (telephon) [inline testsuite/classlibrary/TestPatternProxy.sc]: wouldn't it be good to know which of the two cases fail?

**Comment by telephon** [comment]
well, I should have thought of it in the first place, when I wrote Pbindef, that the argument order matters!

File: testsuite/classlibrary/TestPatternProxy.sc (hunk 1 above)
Comment (brianlheim) [inline testsuite/classlibrary/TestPatternProxy.sc]: i'm confused. what is actually being tested here?

Comment (telephon) [inline testsuite/classlibrary/TestPatternProxy.sc]: we are testing whether the internal structure of the Pbindef matches the one that is specified by the args of the `new` message. For two cases:

1. a fresh Pbindef
2. a Pbindef that has some arguments already

Thinking of it, we could also test if the transformation between
`Pdef(\x, Pbind(\a, 7, \b, 8))`  to `Pbindef(\x, \a, \7, \b, 8)` is consistent in that manner.

Comment (brianlheim) [inline testsuite/classlibrary/TestPatternProxy.sc]: ok, so 1. and 2. are your two test methods :) each test should only test one thing. just for my own sanity, this PR was only designed to fix (2.) right?

Comment (telephon) [inline testsuite/classlibrary/TestPatternProxy.sc]: yes, sure! :)

**Comment by jamshark70** [comment]
Hold on a minute. I need to test something. At first glance, there's a chance that this might break ddwChucklib-livecode.

**Comment by jamshark70** [comment]
I was right. I can't approve these changes.

The provided tests miss a case. That case happens to be critical for me.

```
Pbindef(\test, \a, 1);
Pbindef(\test, \b, 1);
Pbindef(\test).source.pairs
-> [ b, a PatternProxy, a, a PatternProxy ]
```

`\b` now comes before `\a`. This strikes me as counterintuitive. Should `aPbindProxy.set(key, value)` mean that you necessarily want `key` to come first?

It also poses a major problem for ddwChucklib-livecode ("cll"). Of course, you're not under any formal obligation to support quarks, but let me explain anyway. In cll, all data patterns are timed. There is a "default parameter" that determines the process's rhythm. Secondary parameters also associate values with specific metrical positions. If those metrical positions differ from the default parameter's rhythm, the secondary parameters are treated as "sample-and-hold." (I did this independently, but it's the same approach as in TidalCycles.)

I had some problems using Pstep for the sample-and-hold (at one point, I was using some lookahead strategies in some generator objects) -- the problem was that Pstep cues off of the clock time instead of elapsed `\delta` values from the events. So I wrote a PstepDur pattern. This only works if `\dur` is correctly populated. So, every event pattern generated by cll initializes the PbindProxy to put the default parameter first, declared as `[\defaultParmName, \dur], Ptuple(...)`.

It's mandatory that the defaultParm-rhythm pair appears first. Previously, this was guaranteed by initializing the PbindProxy. With the proposed changes, I will have to change my code to do `thePbindProxy.set(pairs[0], pairs[1], \newThing, newPattern)`, or stop using PbindProxy entirely.

**Comment by jamshark70** [comment]
Possible easy fix: If a particular call to PbindProxy:set includes only new keys (no preexisting ones), then the new stuff should be appended, not inserted. I'm not sure that would cover all possible cases, but it would cover mine.

I can imagine a more interesting, possibly more robust way but I need to think about it a bit more.

**Comment by brianlheim** [comment]
oh, i thought that was an intended effect..

why exactly was this a problem in the first place? i've just assumed it was a problem because it's a JITLib issue/PR and yall are the owners in that particular area. but now i guess we've wasted a couple hours because nobody bothered to give a real analysis of the issue.

**Comment by brianlheim** [comment]
(I thought the 'pairs' thing was just a set? so why is anyone depending on order in the first place? is this purely cosmetic?)

**Comment by adcxyz** [comment]
one key may depend on another key being set first (e.g. amp depending on the value of pitch), 
and then the order is important. See e.g. Pkey.help

**Comment by brianlheim** [comment]
ah thanks, that makes sense. sorry to ask such a simple question but is it possible to reorder the keys manually after they've been inserted? if that's possible, does that make the need for this PR go away? again, not sure what the goal is because none was given other than passing a test case

**Comment by telephon** [comment]
no Problem, i just didn't think of that case. I'll fix asap.

**Comment by adcxyz** [comment]
reordering would be really useful, and could just follow the pattern of 
add -> addLast, addFirst, addBefore (otherKey), addAfter (otherKey) 

best a

**Comment by jamshark70** [comment]
One thought I had is that a group of new keys, before an existing key, could be inserted before the existing one.

```
Pbindef(\test, \a, 1);
Pbindef(\test, \b, 1);
// b is new, but attached to no existing key
// so, append: a, b

Pbindef(\test, \c, 1, \b, 1);
// c is new, but "attached" to b, so: a, c, b
```

An analogy to addActions would be more robust, but failing that, a good reason to prefer "insert before known key" is that data dependencies are always in one direction.

I've only half thought through this. Feel free to disregard.

**Comment by telephon** [comment]
Here are two variants that model possible behaviors in a simplified form. Which is better?

https://gist.github.com/telephon/9703b662f234ff9be6fc1497c4554001

**Comment by telephon** [comment]
… here is a suggestion. Feel free to comment!

**Comment by telephon** [comment]
(I'll document once agreed upon)

**Comment by jamshark70** [comment]
If all of the tests pass as written, then I'm fine with it. I haven't thought through any other cases, though.

**Comment by brianlheim** [comment]
can you please put comments in your code or make some attempt to explain it? no idea what `stitchPairs` intends to do

**Comment by telephon** [comment]
don't try yet, it's still not there, sorry for the noise.

**Comment by telephon** [comment]
@brianlheim I'll do. Most important of course is help documentation, I can't explain it in code, I think. For now: it'll conform to the expectations about insertion behavior expressed in this thread :).

**Comment by telephon** [comment]
(see cleaner revision here: #3020)

**Pull request closed without merging**